### PR TITLE
Use (hand-rolled) sparse matrix multiplication

### PR DESCRIPTION
**Pull request by astralblue**
This replaces use of `gonum.org/v1/gonum/mat.(*Dense)` and `gonum.org/v1/gonum/mat.(*VecDense)` with a hand-rolled sparse matrix/vector implementation.  Memory usage is now O(NNZ + R), where NNZ is the number of non-zero entries and R is the number of rows.

### CODE REVIEW

File: pkg/sparse/entry.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+type entrySort []Entry
+
+func (a entrySort) Len() int           { return len(a) }
+func (a entrySort) Less(i, j int) bool { return a[i].Index < a[j].Index }
+func (a entrySort) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func SortEntries(entries []Entry) []Entry {
+	entries = append(entries[:0:0], entries...)
```
Comment (web123pro) [inline pkg/sparse/entry.go]: Could we have a TODO to change this method to do an in-place sort ? If the calling code needs a new copy to be sorted, then it create a copy before calling this function.

File: pkg/basic/server.go [hunk 2]
```diff
@@ -152,7 +182,7 @@ func loadInlineTrustVector(inline *InlineTrustVector) (TrustVector, error) {
 			return nil, errors.Errorf("entry %d: v=%f is out of range (0, inf)",
 				idx, entry.V)
 		}
-		lt.SetVec(entry.I, entry.V)
+		entries = append(entries, sparse.Entry{Index: entry.I, Value: entry.V})
 	}
-	return lt, nil
+	return sparse.NewVector(inline.Size, entries), nil
```
Comment (web123pro) [inline pkg/basic/server.go:197]: At this point in the code, we would have 3 copies of the TrustVector - 1 copy in the InlineTrustVector, 2nd copy in entries slice, and  3rd copy in sparse.NewVector.

File: pkg/sparse/entry.go (hunk 1 above)
Comment (astralblue) [reply pkg/sparse/entry.go]: Fixed! <3

File: pkg/basic/server.go (hunk 2 above)
Comment (astralblue) [reply pkg/basic/server.go:197]: Fixed.  Now we have only two copies at most at any given point in time, and only one copy after this function returns:

- `entries`—discarded at the end of this function
- the constructed `*sparse.Vector`—this is the one that lives on.

**Review by web123pro — APPROVED** [review]

**Pull request merged**